xds/src/test/java/io/grpc/xds/LrsLoadBalancerTest.java

```diff
@@ -0,0 +1,325 @@
+        .updateBalancingState(eq(ConnectivityState.TRANSIENT_FAILURE), pickerCaptor.capture());
+    status = pickerCaptor.getValue().pickSubchannel(mock(PickSubchannelArgs.class)).getStatus();
+    assertThat(status.getDescription()).contains("I failed");
+    assertThat(status.getDescription()).contains("handled by downstream balancer");
+  }
+
+  private void deliverResolvedAddresses(
+      List<EquivalentAddressGroup> addresses, String childPolicy) {
+    PolicySelection childPolicyConfig =
+        new PolicySelection(new FakeLoadBalancerProvider(childPolicy), null, null);
+    LrsConfig config =
+        new LrsConfig(
+            CLUSTER_NAME, EDS_SERVICE_NAME, LRS_SERVER_NAME, TEST_LOCALITY, childPolicyConfig);
+    ResolvedAddresses resolvedAddresses =
+        ResolvedAddresses.newBuilder()
+            .setAddresses(addresses)
+            .setAttributes(
+                Attributes.newBuilder()
+                    .set(XdsAttributes.ATTR_CLUSTER_SERVICE_LOAD_STATS_STORE, loadRecorder)
+                    .build())
+            .setLoadBalancingPolicyConfig(config)
+            .build();
+    loadBalancer.handleResolvedAddresses(resolvedAddresses);
+  }
+
+  private static List<EquivalentAddressGroup> createResolvedBackendAddresses(int n) {
+    List<EquivalentAddressGroup> list = new ArrayList<>();
+    for (int i = 0; i < n; i++) {
+      SocketAddress addr = new FakeSocketAddress("fake-address-" + i);
+      list.add(new EquivalentAddressGroup(addr));
+    }
+    return list;
+  }
+
+  private final class FakeLoadBalancerProvider extends LoadBalancerProvider {
+    private final String policyName;
+
+    FakeLoadBalancerProvider(String policyName) {
+      this.policyName = policyName;
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      LoadBalancer balancer = new FakeLoadBalancer(helper, policyName);
+      childBalancers.add(balancer);
+      return balancer;
+    }
+
+    @Override
+    public boolean isAvailable() {
+      return true;
+    }
+
+    @Override
+    public int getPriority() {
+      return 0;  // doesn't matter
+    }
+
+    @Override
+    public String getPolicyName() {
+      return policyName;
+    }
+  }
+
+  private static final class FakeLoadBalancer extends LoadBalancer {
+    private final Helper helper;
+    private final String name;
+    private boolean shutdown;
+
+    FakeLoadBalancer(Helper helper, String name) {
+      this.helper = helper;
+      this.name = name;
+    }
+
+    @Override
+    public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+      final Subchannel subchannel = new NoopSubchannel(resolvedAddresses.getAddresses());
+      SubchannelPicker picker = new SubchannelPicker() {
+        @Override
+        public PickResult pickSubchannel(PickSubchannelArgs args) {
+          return PickResult.withSubchannel(subchannel);
+        }
+      };
+      helper.updateBalancingState(ConnectivityState.READY, picker);
```

Comment:
This is just testing test implementation. The `subchannel` of the picker and the action of calling `childHelper.updateBalancingState()` should be considered as test input instead of test implementation.

Reply:
This is the implicitly an "input", by my implementation choice: calling `LrsLoadBalancer`'s `handleResolvedAddresses(...)` immediately triggers a subchannel become READY. This merges two steps into one: deliver addresses + deliver subchannel state. We do not care about how subchannels are created, so it's completely up to the fake balancer's implementation.

Reply:
Hiding implicit input in fake lb's `handleResolvedAddresses()` implementation makes the tests very difficult to read and surprises readers: you are testing against a very very special child balancer implementation. 

You can make the input explicit by adding a method `FakeLoadBalancer.deliverBalancingState()`and call it explicitly in the test method - if you don't want to cache the childHelper and call `childHelper. updateBalancingState()` explicitly.

Reply:
How downstream balancer/helper propagate subchannel state/picker up to parent balancer is completely downstream balancer's logic. In this test, its choice is to immediately propagate a READY subchannel when it receives `ResolvedAddresses`.

> makes the tests very difficult to read and surprises readers

Why this would surprise readers? Readers need to know how downstream balancer propagates subchannel state/picker to parent balancer, so it needs to read how this fake downstream balancer is implemented.

> you are testing against a very very special child balancer implementation.

Does this matter to what is being tested and verified? Obviously it does not.

Reply:
> Why this would surprise readers?

A couple of reasons:

- A fake object should normally have a trivial implementation, such as recording the input arguments and constructing a dummy return value. Fake object should normally not implement the overriding methods with any behaviors. Otherwise you should not name it fake object (sometimes I made the same mistake), but give them a meaningful name like `ImmediateReadyLoadBalancer`,  `ImmediateReadyLbProvider`, and `immediate-ready-policy-foo`, and optionally provide javadoc for the behavior. Otherwise readers reading the test methods will ignore/overlook there is a behavior in the fake object implementation.

-  Subchannel turning to READY state is triggered by external event: the transport is connected and notifies the subchannel state listener.  Lb itself can not control when the event is fired.
  Calling `childHelper.updateBalancingState(READY, picker)`explicitly on demand in test methods as a test input is a sound way to simulate that the external event is happening. On the contrary, overriding `handleResolvedAddresses()` to call `childHelper. updateBalancingState(READY, picker)` is a weird behavior as if the lb itself fires the external event automatically.

>Does this matter to what is being tested and verified?

Overriding `handleResolvedAddresses()` this way is technically equivalent to calling a dummy `handleResolvedAddresses()` and immediately calling `childHelper.updateBalancingState(READY, picker)` after it. If all the test methods only need to call the two together, technically there is no difference, but looks all the tests and future tests are relying on this special behavior. Two approaches can technically achieve the same thing but one is hacky.


Just my 2 cents.

Reply:
Sure, sounds solid. Thanks. Updated.